sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -3436,6 +3437,21 @@ class Analyzer(override val catalogManager: CatalogManager) extends RuleExecutor
     }
   }
 
+  /**
+   * Transform UnresolvedBetweenExpression into a [BetweenExpr].
+   */
+  object ResolveBetweenExpression extends Rule[LogicalPlan] {
```

Comment:
We should avoid trying to add new analyzer rules at this point. Could we accomplish this  by combining `UnresolvedBetweenExpression` and `ResolvedBetweeExpression` into just one `BetweenExpression` and make that inherit from `RuntimeReplaceable`? Then its `replacement` can be the `With` expression of interest.

![image](https://github.com/apache/spark/assets/99207096/f96c15a0-bd5a-4870-b13f-2150ff2e58ff)

Reply:
This is what I tried initially. I tried to explain the problem in PR description, but I guess that I wasn't clear enough. Let me try to do this again:
Initially I tried with `BetweenExpr extends RuntimeReplaceable`. The problem that I hit was that `CommonExpressionRef` required dataType of `CommonExpressionDef` to be known for it to be created:
```scala
  def this(exprDef: CommonExpressionDef) = this(exprDef.id, exprDef.dataType, exprDef.nullable)
```
To get around this I would like to create replacement of BetweenExpr after I resolve it's Child expressions and hence after I have it's dataType. I hit a wall doing this so I did this thing with UnresolvedBetween->BetweenExpr (for which I will know the types)->With, which works, but I agree that it would be better if we could avoid this + get rid of extra rule.

I will give it another try. If you guys know a way to get around this I would appreciate some help :)

Reply:
Ok, I hope that i got it right in the latest iteration. Please take a look.

Btw, thanks for the comment, this is definitely much cleaner.